mojo/stdlib/std/os/process.mojo

```diff
@@ -0,0 +1,441 @@
+                fcntl(fd, FcntlCommands.F_GETFD, 0) | FcntlFDFlags.FD_CLOEXEC,
+            )
+            == 0
+        )
+
+    @always_inline
+    fn set_input_only(mut self):
+        """Close the output descriptor/ channel for this side of the pipe."""
+        if self.fd_out:
+            _ = close(rebind[Int](self.fd_out.value()))
+            self.fd_out = None
+
+    @always_inline
+    fn set_output_only(mut self):
+        """Close the input descriptor/ channel for this side of the pipe."""
+        if self.fd_in:
+            _ = close(rebind[Int](self.fd_in.value()))
+            self.fd_in = None
+
+    @always_inline
+    fn write_bytes(mut self, bytes: Span[Byte, _]) raises:
+        """Writes a span of bytes to the pipe.
+
+        Args:
+            bytes: The byte span to write to this pipe.
+
+        Raises:
+            Error: If called on a read-only pipe.
+        """
+        if self.fd_out:
+            self.fd_out.value().write_bytes(bytes)
+        else:
+            raise Error("Can not write from read only side of pipe")
+
+    @always_inline
+    fn read_bytes(mut self, mut buffer: Span[mut=True, Byte]) raises -> UInt:
+        """Read a number of bytes from this pipe.
+
+        Args:
+            buffer: Span[Byte] of length n where to store read bytes. n = number of bytes to read.
+
+        Returns:
+            Actual number of bytes read.
+
+        Raises:
+            Error: If the pipe is in write-only mode.
+        """
+        if self.fd_in:
+            return self.fd_in.value().read_bytes(buffer)
+
+        raise Error("Can not read from write only side of pipe")
+
+
+# ===----------------------------------------------------------------------=== #
+# Process execution
+# ===----------------------------------------------------------------------=== #
+struct Process:
+    """Create and manage child processes from file executables.
+
+    Example usage:
+    ```
+    child_process = Process.run("ls", List[String]("-lha"))
+    if child_process.interrupt():
+        print("Successfully interrupted.")
+    ```
+    """
+
+    var child_pid: c_pid_t
+    """Child process id."""
+
+    var status: Optional[ProcessStatus]
+    """Cached status of the process. `None` if the process has not been waited on yet."""
+
+    fn __init__(out self, child_pid: c_pid_t):
+        """Struct to manage metadata about child process.
+        Use the `run` static method to create new process.
+
+        Args:
+          child_pid: The pid of child processed returned by `posix_spawnp` that the struct will manage.
+        """
+
+        self.child_pid = child_pid
+        self.status = None
+
+    fn __del__(deinit self):
+        """Waits for the process to exit when the `Process` object is destroyed.
+        """
+        try:
+            _ = self.wait()
+        except:
+            # Errors in __del__ should be suppressed.
+            pass
+
+    fn _kill(mut self, signal: Int) -> Bool:
```

Comment:
This is a TOCTOU (time of check, time of use) race I believe (right @barcharcraz?)

Like your comment says, the process can exit between `poll()` and `kill()`, potentially allowing the PID to be reused for a different process. I'd recommend we just call `kill()` directly and check for `ESRCH` errno. The kernel handles this automatically for us:

```mojo
fn _kill(mut self, signal: Int) -> Bool:
  if kill(self.child_pid, signal) == 0:
      return True
  var err = get_errno()
  return err != ESRCH # False if process doesn't exist, True for other errors
```

Reply:
The PID won't be reused even if the process has already exited. The process will become a zombie until reaped because we don't change the default signal disposition for sigchld.

The way SIGCHLD works is confusing, if you set the disposition to SIG_IGN then the process isn't zombified at all (and even the poll might poll the wrong process), but if it's SIG_DFL then you _do_ get zombies even though the "default" action is to ignore the signal.

Anyway, since we aren't messing with SIGCHLD we do get zombies. Either we poll and the process has exited, in which case the poll will reap the process and we never kill, or it hasn't and the process isn't reaped. After being killed it will be zombified to be handled by a later `wait`/`pool`/`__del__`